server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -630,6 +634,12 @@ private void balanceByWeights(NodeSorter sorter) {
                             sorter.sort(0, relevantNodes);
                             lowIdx = 0;
                             highIdx = relevantNodes - 1;
+                            if (allocation.hasTargetRelocatingShards()) {
+                                // The shard may have been relocated on the ModelNode, but not on the cluster (RoutingNodes) if
+                                // it is throttled. The `hasTargetRelocatingShards` check does not account for this case.
+                                // This might be a non-issue since our intention is to have no Throttle with early returns.
```

Comment:
The supposed contract of tryRelocateShard, according to the method comment, is that true means the move was executed "on the cluster" 🤔 Pre-existing outdated comment, seems like.

How about we update this code https://github.com/elastic/elasticsearch/blob/7d6e1e4d84542d8b3d7bbceef6c3d0f9db25727e/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java#L1152-L1155 to assert a non-throttling answer and always do the routingNodes relocate call? With your complete code changes for this task, I would never expect the balancer to run into THROTTLE again -- we're trying to avoid THROTTLE situations, which can lead to overriding not-preferred answers.

Reply:
Yeah added such an assertion in the caller, see dbd0fede791b9a31aa750108d32f7d9b99775d18